subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.java

```diff
@@ -0,0 +1,145 @@
+ * to attempt to locate a matching role by its usage characteristics.
+ *
+ * @since 8.0
+ */
+public enum ConfigurationRoles implements ConfigurationRole {
+    /**
+     * An unrestricted configuration, which can be used for any purpose.
+     *
+     * This is available for backwards compatibility, but should not be used for new configurations.  It is
+     * the default role for configurations created when another more specific role is <strong>not</strong> specified.
+     */
+    @Deprecated
+    LEGACY(true, true, true, false, false, false),
+
+    /**
+     * Meant to be used only for consumption by other projects.
+     */
+    INTENDED_CONSUMABLE(true, false, false, false, false, false),
+
+    /**d
+     * Meant to be used only for resolving dependencies.
+     */
+    INTENDED_RESOLVABLE(false, true, false, false, false, false),
+
+    /**
+     * Meant to be used only for declaring dependencies.
+     *
+     * AKA {@code INTENDED_DECLARABLE}.
+     */
+    INTENDED_BUCKET(false, false, true, false, false, false),
+
+    /**
+     * Meant to be used only for consumption, permits other usage but emits warnings if used otherwise.
+     */
+    @Deprecated
+    DEPRECATED_CONSUMABLE(true, true, true, false, true, true),
+
+    /**
+     * Meant to be used only for resolution, permits other usage but emits warnings if used otherwise.
+     */
+    @Deprecated
+    DEPRECATED_RESOLVABLE(true, true, true, true, false, true);
+
+    private final boolean consumable;
+    private final boolean resolvable;
+    private final boolean declarableAgainst;
+    private final boolean consumptionDeprecated;
+    private final boolean resolutionDeprecated;
+    private final boolean declarationAgainstDeprecated;
+
+    /**
+     * Locates a pre-defined role allowing the given usage.
+     *
+     * @param consumable whether this role is consumable
+     * @param resolvable whether this role is resolvable
+     * @param declarableAgainst whether this role is declarable against
+     * @param consumptionDeprecated whether this role is deprecated for consumption
+     * @param resolutionDeprecated whether this role is deprecated for resolution
+     * @param declarationAgainstDeprecated whether this role is deprecated for declaration against
+     *
+     * @return the role enum token with matching usage characteristics, if one exists; otherwise {@link Optional#empty()}
+     */
+    public static Optional<ConfigurationRoles> byUsage(boolean consumable, boolean resolvable, boolean declarableAgainst, boolean consumptionDeprecated, boolean resolutionDeprecated, boolean declarationAgainstDeprecated) {
+        for (ConfigurationRoles role : values()) {
+            if (role.consumable == consumable && role.resolvable == resolvable && role.declarableAgainst == declarableAgainst && role.consumptionDeprecated == consumptionDeprecated && role.resolutionDeprecated == resolutionDeprecated && role.declarationAgainstDeprecated == declarationAgainstDeprecated) {
```

Comment:
Another argument for implementing a `DefaultConfigurationRole` -- this should be the `equals` method.

Reply:
I don't think this method or an equals method would be needed at all with a default role implementation class.  But I think other concerns ought to decide this.